src/Emogrifier/HtmlProcessor/AbstractHtmlProcessor.php

```diff
@@ -84,11 +91,27 @@ public function render()
      */
     public function renderBodyContent()
     {
-        $bodyNodeHtml = $this->domDocument->saveHTML($this->getBodyElement());
+        $bodyNodeHtml = $this->renderDomDocument($this->getBodyElement());
 
         return \str_replace(['<body>', '</body>'], '', $bodyNodeHtml);
     }
 
+    /**
+     * Renders the DOMDocument as HTML, eliminating any invalid closing tags for void elements.
+     *
+     * @param \DOMNode|null $node Optional parameter to output a subset of the document.
+     *
+     * @return string
+     */
+    protected function renderDomDocument(\DOMNode $node = null)
```

Comment:
Please let's split this into two functions instead of having kind of a feature switch with the parameter (which is a pattern to be avoided). Or maybe we can move the `preg_replace` into a private method that is called both from `renderBodyContent` and `render`.

Reply:
The method as proposed is a 'fixed' version of [`DOMDocument::saveHTML()`](https://secure.php.net/manual/en/domdocument.savehtml.php) which has the same method signature and feature switch parameter.  The regex replacement could be moved to a separate method such as `removeSelfClosingTagsClosingTags()`, but without `renderDomDocument()`, both `render` and `renderBodyContent` would have to call `removeSelfClosingTagsClosingTags()` on the result of `DOMDocument::saveHTML()` which to me seems like a duplication, albeit a very small one.  What do you think?

Reply:
Some "structural duplication" is okay, I think. Generally, I prefer this approach:

```
function a1() {
  doSomething;
  sharedParts();
}

function a2() {
  doSomethingElse;
  sharedParts();
}

function sharedParts() {
  …
}
```

over this:

```
function a($someSwitch) {
  if ($someSwitch) {
    doSomething;
  } else {
    doSomethingElse;
  }

  sharedStuff;
}
```